packages/fetchai/connections/p2p_libp2p_client/connection.py

```diff
@@ -0,0 +1,314 @@
+
+        # delegates certificates
+        self.delegate_certs = list(libp2p_delegate_certs)
+
+        # select a delegate
+        index = random.randint(0, len(self.delegate_uris) - 1)
+        self.node_uri = self.delegate_uris[index]
+        # self.node_cert = self.delegate_certs[index]
+
+        # tcp connection
+        self._reader = None  # type: Optional[asyncio.StreamReader]
+        self._writer = None  # type: Optional[asyncio.StreamWriter]
+
+        self._loop = None  # type: Optional[AbstractEventLoop]
+        self._in_queue = None  # type: Optional[asyncio.Queue]
+        self._process_message_task = None  # type: Union[asyncio.Future, None]
+
+    async def connect(self) -> None:
+        """
+        Set up the connection.
+
+        :return: None
+        """
+        if self.connection_status.is_connected:
+            return
+        if self._loop is None:
+            self._loop = asyncio.get_event_loop()
+        try:
+            # connect libp2p client
+            self.connection_status.is_connecting = True
+
+            # connect the tcp socket
+            self._reader, self._writer = await asyncio.open_connection(
+                self.node_uri.host, self.node_uri._port, loop=self._loop
+            )
+
+            # send agent address to node
+            await self._setup_connection()
+
+            self.connection_status.is_connecting = False
+            self.connection_status.is_connected = True
+
+            # start receiving msgs
+            self._in_queue = asyncio.Queue()
+            self._process_messages_task = asyncio.ensure_future(
+                self._process_messages(), loop=self._loop
+            )
+        except (CancelledError, Exception) as e:
+            self.connection_status.is_connected = False
+            raise e
+
+    async def _setup_connection(self):
+        await self._send(bytes(self.agent_addr, "utf-8"))
+        await self._receive()
+
+    async def disconnect(self) -> None:
+        """
+        Disconnect from the channel.
+
+        :return: None
+        """
+        assert (
+            self.connection_status.is_connected or self.connection_status.is_connecting
+        ), "Call connect before disconnect."
+        self.connection_status.is_connected = False
+        self.connection_status.is_connecting = False
+
+        assert self._process_messages_task is not None
+        assert self._writer is not None
+
+        if self._process_messages_task is not None:
+            self._process_messages_task.cancel()
+            # TOFIX(LR) mypy issue https://github.com/python/mypy/issues/8546
+            # self._process_messages_task = None
+
+        logger.debug("disconnecting libp2p client connection...")
+        self._writer.write_eof()
+        await self._writer.drain()
+        self._writer.close()
+        # TOFIX(LR) requires python 3.7 minimum
+        # await self._writer.wait_closed()
```

Comment:
Can you change it to a python 3.6 compatible api please

Reply:
It is. Commented code requires python 3.7